src/OrchardCore.Modules/Orchard.Lucene/LuceneIndexingService.cs

```diff
@@ -50,80 +51,88 @@ public class LuceneIndexingService
 
         public async Task ProcessContentItemsAsync()
         {
-            
-            // TODO: Lock over the filesystem
+            // TODO: Lock over the filesystem in case two instances get a command to rebuild the index concurrently.
 
-            var allIndexes = new Dictionary<string, int>();
+            var allIndices = new Dictionary<string, int>();
 
             // Find the lowest task id to process
             int lastTaskId = int.MaxValue;
             foreach (var indexName in _indexManager.List())
             {
                 var taskId = _indexingState.GetLastTaskId(indexName);
                 lastTaskId = Math.Min(lastTaskId, taskId);
-                allIndexes.Add(indexName, taskId);
+                allIndices.Add(indexName, taskId);
             }
 
-            if (!allIndexes.Any())
+            if (!allIndices.Any())
             {
                 return;
             }
 
             IEnumerable<IndexingTask> batch;
+            var shellContext = _shellHost.GetOrCreateShellContext(_shellSettings);
 
             do
             {
-                // Load the next batch of tasks
-                batch = (await _indexTaskManager.GetIndexingTasksAsync(lastTaskId, BatchSize)).ToArray();
-
-                if (!batch.Any())
+                // Create a scope for the content manager
+                using (var scope = shellContext.CreateServiceScope())
                 {
-                    break;
-                }
+                    // Load the next batch of tasks
+                    batch = (await _indexingTaskManager.GetIndexingTasksAsync(lastTaskId, BatchSize)).ToArray();
 
-                foreach (var task in batch)
-                {
-                    foreach (var index in allIndexes)
+                    if (!batch.Any())
                     {
-                        // TODO: ignore if this index is not configured for the content type
-
-                        if (index.Value < task.Id)
-                        {
-                            _indexManager.DeleteDocuments(index.Key, new string[] { task.ContentItemId });
-                        }
+                        break;
                     }
 
-                    if (task.Type == IndexingTaskTypes.Update)
+                    foreach (var task in batch)
                     {
-                        var contentItem = await _contentManager.GetAsync(task.ContentItemId);
-                        var context = new BuildIndexContext(new DocumentIndex(task.ContentItemId), contentItem, contentItem.ContentType);
-
-                        // Update the document from the index if its lastIndexId is smaller than the current task id. 
-                        await _indexHandlers.InvokeAsync(x => x.BuildIndexAsync(context), Logger);
+                        var contentManager = scope.ServiceProvider.GetRequiredService<IContentManager>();
```

Comment:
Yes, it's on purpose. It's a batch of content items, using a custom scope will dispose it correctly, including the Isession object. This way we don't keep the content items in memory if we need to iterate over millions of them.